### PR TITLE
Update docs

### PR DESCRIPTION
**Pull request by joecummings**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add unit tests for any new functionality
- [ ] update docstrings for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
	- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

File: docs/source/api_ref_models.rst [hunk 1]
```diff
@@ -2,105 +2,409 @@
...
-All models from the `Llama3 family <https://llama.meta.com/llama3/>`_.
+.. currentmodule:: torchtune.models.llama3
+
+`Llama 3 <https://llama.meta.com/llama3/>`_ - released in April of 2024 - is an auto-regressive language model that uses an optimized transformer architecture.
+The tuned versions use supervised fine-tuning (SFT) and reinforcement learning with human feedback (RLHF) to align with human preferences for helpfulness and safety.
+The model was pretrained on over **15 trillion tokens** of data from publicly available sources. The pretraining data has a cutoff of March 2023 for the 8B and December 2023
+for the 70B models respectively. The fine-tuning data includes publicly available instruction datasets, as well as over **10M human-annotated examples**.
+You can read more about the model from the official `model card <https://github.com/meta-llama/llama3/blob/main/MODEL_CARD.md#model-details>`_.
+
+Download the model
```
Comment (SalmanMohammadi) [inline docs/source/api_ref_models.rst:19]: I think we've sufficiently covered downloading models in our other docs

File: docs/source/api_ref_models.rst [hunk 2]
```diff
@@ -2,105 +2,409 @@
...
+^^^^^^^^^^^^^^^^^^
+
+Pre-trained models can be downloaded from the Hugging Face Hub with the following command:
 
 .. code-block:: bash
 
-    tune download meta-llama/Meta-Llama-3-8B-Instruct --hf-token <ACCESS_TOKEN>
+    tune download meta-llama/Meta-Llama-3-8B-Instruct --output-dir /tmp/Meta-Llama-3-8B-Instruct --hf-token <ACCESS_TOKEN>
+
+You can find all model checkpoints on the `official Meta Llama model page <https://huggingface.co/collections/meta-llama/meta-llama-3-66214712577ca38149ebb2b6>`_.
```
Comment (SalmanMohammadi) [inline docs/source/api_ref_models.rst:28]: I like including this for every model family though

File: docs/source/api_ref_models.rst [hunk 3]
```diff
@@ -2,105 +2,409 @@
...
-    llama3.llama3_70b
-    llama3.lora_llama3_8b
-    llama3.qlora_llama3_8b
-    llama3.lora_llama3_70b
+    llama3_tokenizer
+
+Usage
+^^^^^
+
+You can specify your Llama3 model through code ...
```
Comment (SalmanMohammadi) [inline docs/source/api_ref_models.rst:67]: can probably just link to somewhere else in our docs for this

File: docs/source/api_ref_models.rst [hunk 4]
```diff
@@ -2,105 +2,409 @@
...
 .. autosummary::
     :toctree: generated/
     :nosignatures:
 
-    llama3.llama3_8b
-    llama3.llama3_70b
-    llama3.lora_llama3_8b
-    llama3.qlora_llama3_8b
-    llama3.lora_llama3_70b
+    llama3_tokenizer
```
Comment (SalmanMohammadi) [inline docs/source/api_ref_models.rst:62]: worth linking to something like https://github.com/meta-llama/llama3/blob/main/llama/tokenizer.py for a one-file implementation?

**Comment by SalmanMohammadi** [comment]
Overall this is a huge improvement. I'm not married to the blurbs - we can link to model cards/papers which sufficiently explain the models IMO.

I'd love to see the following things (some of which are here already):

- A top-level note to let users know they can swap all the models they see here across all of our recipes. How much detail we provide on doing so vs. linking to other docs I'm open to discussing.
- **Memory usage of different models** - I think this is just a nice QOL add, even just to say the memory usage default for bf16.
- The explainer on generic builders has been sorely missing. I wouldn't mind a little example showing when to use one (e.g. for tinyllama or something)
- Would it be inappropriate to immortalize a contributor's contribution if they contributed a specific model, by linking to their contribution in the model section?
- I think the tokenizer sections could use a little love, particularly since we've fleshed these out a bit. For example, worth specifying llama tokenizer is a sentencepiece tokenizer, and not the fast BPE tokenizer?

More thoughts maybe to come. Thanks for reviving this.

File: docs/source/api_ref_models.rst [hunk 5]
```diff
@@ -2,105 +2,409 @@
...
+    :toctree: generated/
+    :nosignatures:
+
+    llama3_8b
+    lora_llama3_8b
+    qlora_llama3_8b
+    llama3_70b
+    lora_llama3_70b
+
+If you want to experiment with different sizes of models or train a model from scratch, we also provide generic builder
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply docs/source/api_ref_models.rst:45]: and uh, omit the train from scratch?

File: docs/source/api_ref_models.rst [hunk 6]
```diff
@@ -2,105 +2,409 @@
...
-llama3
+This page contains information on all model architectures available in the core torchtune library.
+
+Llama3
 ------
 
-All models from the `Llama3 family <https://llama.meta.com/llama3/>`_.
+.. currentmodule:: torchtune.models.llama3
+
+`Llama 3 <https://llama.meta.com/llama3/>`_ - released in April of 2024 - is an auto-regressive language model that uses an optimized transformer architecture.
```
Comment (SalmanMohammadi) [inline docs/source/api_ref_models.rst:13]: I kind of see what you mean about maintaining the blurbs now, we could be writing a lot of these

**Comment by SalmanMohammadi** [comment]
Also, maybe this is obvious to everyone, but it might be a good opportunity to add a little noobsplainer™  about what's up with all our `lora` model and component builders. We can defer to the tutorial for in-depth details here. For example, `lora_llama2_7b` isn't actually another Llama2 model you can download or find the weights for, it's just the LoRA version of the same model, and here's how to specify it, et cetera. Maybe my section in the memory glossary is a good link here?

File: docs/source/api_ref_models.rst (hunk 6 above)
Comment (felipemello1) [reply docs/source/api_ref_models.rst:13]: The description is interesting, but we dont follow the same standard for the other models. I think that giving general info, e.g. paper link, size, config link may be good enough.

File: docs/source/api_ref_models.rst [hunk 7]
```diff
@@ -2,105 +2,409 @@
...
-    llama3.lora_llama3_70b
+    llama3_tokenizer
+
+Usage
+^^^^^
+
+You can specify your Llama3 model through code ...
+
+.. code-block:: python
+
```
[the remark this answers is not in the document]
Comment (felipemello1) [reply docs/source/api_ref_models.rst:70]: Maybe this could be in one of the tutorials, e.g. all about configs: 
Or a new tutorial: "using torchtune with notebooks"

**Review by felipemello1 — COMMENTED** [review]
I like it, but I think it needs a bit more of brainstorming. I worry a bit with noise/information ratio. I feel that maybe reorganizing some information/tutorials can yield a better result (but also more work).

For example: 
- should we have a single page with all models, or a page per models with dropdown? I remember that i couldnt find how to download gemma in torchtune page. API makes me think of code, and not information about models.
- Should we share how to download/instantiate a model, but not how to finetune? Is there a better approach?
- If/when we have model results memory/tokens per second, should we add it here too?
- Is this the right place to expose the tokenizer / builder? Or is this too low level?

File: docs/source/api_ref_models.rst [hunk 8]
```diff
@@ -2,105 +2,409 @@
...
+Download the model
+^^^^^^^^^^^^^^^^^^
+
+Pre-trained models can be downloaded from the Hugging Face Hub with the following command:
 
 .. code-block:: bash
 
-    tune download meta-llama/Meta-Llama-3-8B-Instruct --hf-token <ACCESS_TOKEN>
+    tune download meta-llama/Meta-Llama-3-8B-Instruct --output-dir /tmp/Meta-Llama-3-8B-Instruct --hf-token <ACCESS_TOKEN>
```
Comment (felipemello1) [inline docs/source/api_ref_models.rst:26]: what if we replace this "download" part, with a recommendation to read the config docstring? When I was trying torchtune, it was not obvious to me that the details would be there.

File: docs/source/api_ref_models.rst (hunk 1 above)
Comment (SalmanMohammadi) [reply docs/source/api_ref_models.rst:19]: Having slept on it, I actually like this, particularly if we make the move towards a single page per model family. I think it adds value for users who maybe haven't immediately clocked the usage of `ignore-patterns` or just generally want to get started with fine-tuning a given model without thinking too hard about it.
+1

File: docs/source/api_ref_models.rst (hunk 6 above)
Comment (ebsmothers) [reply docs/source/api_ref_models.rst:13]: Personally I would support adding a quick 1-2 sentences at the top for each model though. I think this helps give a quick idea of what the main points are, then can point to references (imo this version has about 25% too much info though, keep it short and sweet). Agree with @felipemello1's point.. whatever precedent we set we should follow it for every model.

File: docs/source/api_ref_models.rst (hunk 1 above)
Comment (ebsmothers) [reply docs/source/api_ref_models.rst:19]: > if we make the move towards a single page per model family

Where are we at on this? I think this would be a nice thing to do

**Comment by ebsmothers** [comment]
Overall I think this is a huge improvement, we just need to ensure the updates are applied consistently across all the different models we support. A few responses to various comments I've seen so far:

(1) I see a few comments about noise/information ratio, especially around which APIs we do and don't document. Imo we shouldn't compromise on documenting our public APIs, even if it means there is extra stuff on our docs pages. Instead, we should structure our docs such that the important stuff is easy to find, and the less pressing stuff is still available. I think separating models out into separate pages will help, as will structuring each individual page from most essential info at the top to least essential info at the bottom. 

(2) Re stuff like commands to fine-tune a given model or a given model's memory, I don't think we should put it here. Instead, we should provide this information elsewhere and link to it. These are API docs and so first and foremost their job is to provide everything a user needs to know about interacting with said APIs. So the existing example usage is the right idea imo -- it shows how to use the provided APIs in Python in a minimal and self-contained way. (Though I would think about how we can make the example usage even a bit clearer, the checkpointer portion is unfortunately a bit of a black box in its current format.)

> Would it be inappropriate to immortalize a contributor's contribution if they contributed a specific model, by linking to their contribution in the model section?

This I wouldn't do personally. Unlike tutorials where we can reference a contributor in a single line at the top, our API docs shouldn't have any extra fluff
